### ISSUE TITLE
pre { tab-size: 4 }

### ISSUE THREAD

**Issue by Rich-Harris** [isaacs/github#170]
For some unfathomable reason, browsers render tabs as 8 spaces. Even among the minority of developers who agree that [tabs > spaces](http://lea.verou.me/2012/01/why-tabs-are-clearly-superior/), I've yet to find anyone who thinks that makes sense.

Happily, [most browsers](http://caniuse.com/#search=tab-size) allow you to set the tab width with the CSS `tab-size` property. Adding something like the following to the CSS used on github.com and gist.github.com would make code previews much more readable:

``` css
pre {
  -moz-tab-size: 4;
  tab-size: 4;
}
```

**Comment by flcoder** [comment]
:+1: for .editorconfig

**Comment by dfcreative** [comment]
Wow. So many +1s and still nothing.
Wish I could make for it somehow, all my projects use tabs.

**Comment by lxe** [comment]
I think the 8-character wide tabs are appropriate by default, considering Linus Torvalds uses 8-character wide tabs for git and Linux. Might as well adhere to git style guide. I like the `.editorconfig` approach though.

**Comment by dfcreative** [comment]
I believe that [Paving the Cowpaths](http://designingsocialinterfaces.com/patterns/Pave_the_Cowpaths) approach would work the best in that case. Just find out how many spaces do ppl use to indent [in a target syntax] and choose the proper default.

**Comment by lorents** [comment]
:+1: for .editorconfig

**Comment by RReverser** [comment]
@flcoder @lxe @lorents See above for temp solution: https://github.com/RReverser/github-editorconfig

**Comment by seanCodes** [comment]
+1 for sticky `ts`. Any luck @bkeepers?
Also +1 for native `.editorconfig` support (without plugin).

**Comment by lucasvickers** [comment]
+1 configurable.  We shouldn't have to install extensions to make your a paid service readable (some of us pay anyway)

**Comment by r10s** [comment]
.editorconfig sounds like a good solution, esp. in addition to a simple (and persistent) user option :-)

**Comment by sha-odoo** [comment]
:+1:  for `ts`

**Comment by duncan** [comment]
For the longest time, I didn't care about tab spacing. Then I started using Go and `go fmt` enforces tabs. /sigh/ So, another vote for a solution that sticks per user or uses `.editorconfig` or something...

**Comment by sapics** [comment]
+1 for .editorconfig

**Comment by mdax** [comment]
+1 for tab-size config in profile or per repo

**Comment by mdax** [comment]
none of the solutions above work for all views, like pull request view.  simple tab-size preference setting would be phenomenal

**Comment by alexmingoia** [comment]
It's unfortunate that the most popular webapp for sharing code doesn't support simple tab preferences.

Please add a tab-size preference in the settings.

**Comment by photodude** [comment]
:+1:  :100:  for .editorconfig
support some basic user and/or repo preferences like tabs.

**Comment by alexmingoia** [comment]
Reading it from `.editorconfig` does not make sense. The whole point of tabs is so that I can set my own preference to view everything.

**Comment by alganet** [comment]
@alexmingoia this isn't necessarily true for languages that both require tabs and are sensible to indentation. Shell script is a popular one: you can [indent heredoc code blocks using tabs, but not spaces](http://en.wikipedia.org/wiki/Here_document#Unix_shells). That forces me to use 8-space indentations on shell files I intent to publish on GitHub, which sucks a lot for readers.

**Comment by abyss-outpost** [comment]
+1

**Comment by david0178418** [comment]
+1 for sticky ts and +1 for native .editorconfig support.  The idea that a tab should be 8 spaces wide is largely held by those waging the great Spaces-vs-Tabs War and waving the Spaces banner.

How many more do we need to get this onto someones priority list, @bkeepers?  It seems to have died since this issue was created over a year ago.

**Comment by bathos** [comment]
@david0178418 agreed -- it's 8 (and not, say, 13) because it's the largest number Big Space could get away with while maintaining a thin veneer of plausible deniability re: having taken punitive action against the infidels. That's my theory anyway. :)

**Comment by avesse** [comment]
+1 please. Don't care if it's .editorconfig or not, but it should be settable per project.

**Comment by WORMSS** [comment]
I would be interested to see if you did change it to 4 by default, how many people would actually complain.

**Comment by stevstrong** [comment]
Please please please
make tab size configurable!

**Comment by bryc** [comment]
Please do something about the tabs. I'm sure the 8 tab width is nice for some, but I use 4 tab width, and it makes some lines longer, the more indented they are. Messes up the view of the code, and I can't predict what ultimately will be displayed!

There should be an option to configure tab width.

**Comment by bdpdx** [comment]
+1 for customizable, persistent tabs.  Please do something @github.

**Comment by gregory-yet** [comment]
:+1: for .editorconfig

**Comment by aaronduchateau** [comment]
+1

**Comment by Alhadis** [comment]
Nobody should be having to change anything on their end just to browse a repo's code properly. This should be handled by the repo's owner/organisation, because *only* the people maintaining the project know what's best for their code's readability. "Standard" tab-sizes be damned, it's because of wilful negligence of people's individual coding preferences that people are being "forced" into switching to spaces.

(Please add this, Github)[http://i.imgur.com/h3z2zUb.png]. I'm not joking.

**Comment by Nate-Wilkins** [comment]
^^^

**Comment by Rich-Harris** [comment]
Surely the whole *point* of tabs is that readability is in the eye of the beholder? A lot of people prefer 2-space indentation; I much prefer 4. Setting tab width on a per-project basis rather than per-user defeats the object. (Also, remember that 99% of repos won't bother with this setting or with an .editorconfig file! The burden shouldn't be on repo owners.)

**Comment by lucasvickers** [comment]
Except most projects that have coding guidelines (i.e. most large ones)

**Comment by Alhadis** [comment]
> The burden shouldn't be on repo owners.
Actively-maintained repositories will be taken care of. GitHub could even allow users to globally set the tab settings of *all* their repos from their profile page (because they'd surely know many users have hundreds of repos).

> Except most projects that have coding guidelines (i.e. most large ones)
Exactly. Furthermore, it's ridiculous that projects that enforce tabs as their coding style are left to have it looking like crap unless somebody has a custom extension installed just to tweak this behaviour. Which, by the way, really shouldn't have to exist. Yet it does.

**Comment by Alhadis** [comment]
[Submitted support request to GitHub concerning this](http://pastebin.com/FxqYJkct).

**Comment by Rich-Harris** [comment]
Thanks @Alhadis. Hope you get more of a response than I did. As the originator of the issue though I have to take issue with this sentiment:

> The only people who should decide how their code is presented to the reader are the ones maintaining it. Period.

I opened this issue because the purpose of tabs is to allow people to read and write code in the way that maximises readability *for them*. It's the only way to end silly arguments about whether tabs should be equivalent to 2 spaces or 4 spaces. If a project really can't abide the idea of people choosing what tab width works for them, they should be using spaces, not tabs.

**Comment by lucasvickers** [comment]
Debating project vs personal tab spacing is irrelevant.  Given people have different desires/needs, it sounds like github should provide something flexible.  i.e. global project settings and the possibility for a user override.

**Comment by Alhadis** [comment]
@Rich-Harris: As an arrogant 4-character-wide tab-lover, I'd be inclined to agree with you. However, the "whitespace war" is an ongoing fire that'll never burn out. Like the "tomato sauce in the fridge" or the "toilet roll facing the wall" arguments, it's an inconvenience that's sustained on aesthetic grounds that can be justified either way. Much as I'd love to see the whole world using tabs for what they're designed for, I know people's formatting preferences will inevitably differ.

>  and the possibility for a user override.

No. See, that's just as bad. People have different interpretations of what constitutes "clean indentation", but a code's author obviously indents code with particular alignments in mind. If a coder's hell-bent on having their code displayed with a specific tab-size, they'll resort to using spaces to ensure readers see their work the way they obviously want them to.

Giving users an option to ignore the author's intent is killing off half the purpose of preserving a project's consistency.

**Comment by lucasvickers** [comment]
So then what's your solution to the whitespace war?

**Comment by Alhadis** [comment]
@lucasvickers: Neither side will win. But there is a truce to be found in `.editorconfig`. Developers will always have their personal preferences (and they'll inevitably find ways to bitch about why each is superior), but with editorconfig files alleviating many of the more practical arguments for/against tabs, hopefully the whitespace-war will be a lot less insufferable.

**Comment by Rich-Harris** [comment]
Don't know when this happened, but it looks like @github have now added support for .editorconfig

**Comment by bardware** [comment]
I just stumbled upon this when I created a gist and set indentation to tabs instead of spaces. There is a second dropdown that does not seem to be saved when `tab` is set. It allows to set width to 2, 4 or 8 spaces.

**Comment by Alhadis** [comment]
Interesting. This came in good timing with the reply I received from GitHub's support department several days ago:

> Hey John,
>
> We definitely appreciate the feedback. We've added a note to this suggestion on our internal Feature Request List. No promises if we might add something to support what you're describing, but we > appreciate the input.
>
> Thanks.

I guess it's a start...

**Comment by Xyrak** [comment]
+1 for whatever option, either:
a file (.editorconfig)
?ts=4 only once
a gui option per repository

But please add something.

**Comment by joetemp** [comment]
Please... can we please edit tab size? Thanks.

**Comment by zimme** [comment]
Having GitHub respecting a `.editorconfig` file in the repo would be pretty nice actually.

**Comment by Alhadis** [comment]
> Edit: My bad, already implemented.

@zimme : Where?

**Comment by zimme** [comment]
@Alhadis: When you have a .editorconfig in your repository it will respect it when viewing code on GitHub.

`indent_style = tab` and `indent_size = 4` shows tabs with 4 columns instead of 8

**Comment by Alhadis** [comment]
Oh! Damn, how long has that been there for? I guess I never noticed because I had [this](https://github.com/sindresorhus/github-tab-size) installed.

Thank GOD for that.

**Comment by WORMSS** [comment]
@Alhadis I've kept using the chrome extension. It's better than relying on a project to have a .editorconfig file set up correctly.

**Comment by Alhadis** [comment]
@WORMSS Yeah, ditto, man, shame this hasn't caught on with more people yet... :\

**Comment by benatkin** [comment]
Sweet, it works! **Finally!** It doesn't seem to work for diffs though.

**Comment by pffy** [comment]
works! peachy keen!! thanks :+1:

**Comment by gatzkerob** [comment]
All these workarounds are good for our eyes only, but I'm planning to send a link to a potential employer. I don't want them to be turned off by line after line of misaligned code, and "?ts=4/8" seems a little clumsy. If no one can agree on what the universal tab size should be, then perhaps it would be better to let us set it on an individual level from the settings menu.

**Comment by Alhadis** [comment]
@gatzkerob You can drop an `.editorconfig` file in the repo to set the tab-size display. =)

**Comment by adrianblynch** [comment]
@Alhadis - I'm not seeing any change when viewing [code for this repo](https://github.com/adrianblynch/hapi-everything).

Does my [.editorconfig](https://github.com/adrianblynch/hapi-everything/blob/master/.editorconfig) looks OK?

I still see [8 space tabs here](https://github.com/adrianblynch/hapi-everything/blob/master/index.js).

**Comment by Alhadis** [comment]
@adrianblynch : You need to tell `.editorconfig` that those rules are global (as in, they apply to every file-type and file-format). You do that by adding `[*]` before the global rules. 

So, the contents of your `.editorconfig` should be this:
```
root = true

[*]
indent_style = tab
indent_size = 4

```

**Comment by adrianblynch** [comment]
@Alhadis - Perfect! Thanks :dancers: 

https://github.com/adrianblynch/hapi-everything/blob/master/index.js

**Comment by Alhadis** [comment]
Looks great. :+1:

**Comment by marcomorain** [comment]
+1

**Comment by igoradamenko** [comment]
:+1:

**Comment by danielbayley** [comment]
Is `.editorconfig` supposed to work for gists?

**Comment by jimmywarting** [comment]
It would also be great if i can set a fallback .editorconfig if such file doesn't exist in the repo.
I think i would also like to override others .editorconfig file cuz i prefer tab size=4 and tab instead of space

**Comment by jimmywarting** [comment]
Another question, can you not put this .editorconfig file in the .github folder?
Don't like to clutter down my root folder with bunch of stuff that don't concern my environment. I only put in my repo since it solve viewing code on github

**Comment by WORMSS** [comment]
@jimmywarting https://chrome.google.com/webstore/detail/tab-size-on-github/ofjbgncegkdemndciafljngjbdpfmbkn

https://github.com/sindresorhus/tab-size-on-github

Much easier than waiting for github to fix it.

**Comment by tinco** [comment]
So now that github supports `.editorconfig`, can the default tab size finally be reduced to 3 or 4? If it's worth anything `ruby/ruby` sports one. Besides that it was a ridiculous argument anyway, any fixed tab size will 'break' any project that uses mixed tabs and spaces and depends on tabs to be any fixed width that's not whatever they have as a default, mixing tabs and spaces is a broken idea to begin with.

Github is actively hurting the tabs vs spaces debate by making tabs look ridiculous. I bet there's projects out there that use spaces for indentation purely because the author feels tabs look stupid on Github.

**Comment by Alhadis** [comment]
> Github is actively hurting the tabs vs spaces debate by making tabs look ridiculous. I bet there's projects out there that use spaces for indentation purely because the author feels tabs look stupid on Github.

This.

I'm actually partially-convinced GitHub aren't bothering to do anything about this issue because they'd prefer the entire world use spaces instead of tabs.

**Comment by methodbox** [comment]
Agreed. I see no point in the conveniences modern text editors provide (like setting your tab width to 2 or 4 spaces so you don't have to mash the spacebar all day) if those tabs are then intentionally converted to look like #@$%.

The odd thing to me is that viewing in the web console my spacing is respected; why not on Github?

I can even force tabs=spaces and set an indent level in Sublime (which has no effect on what I see in Sublime) for setting output for rich text editors (mostly) and yet the spacing remains horrible once I push.

Thanks for the work you guys have done. ```.editorconfig``` is an easy enough fix for me.

**Comment by ORESoftware** [comment]
ummm, so what is the definitive solution for this?

**Comment by methodbox** [comment]
@ORESoftware
Just add a file to your repo called .editorconfig and set the tab values there.

Here's an example from my NMESH repo:

https://github.com/methodbox/nmesh/blob/master/.editorconfig

**Comment by danielbayley** [comment]
While `.editorconfig` is a step in the right direction, a per repository (and/or per user?) setting would be _much_ better.

**Comment by Alhadis** [comment]
> While .editorconfig is a step in the right direction, a per repository (and/or per user?) setting would be much better. Is this being considered?

I've been imploring GitHub to do this for years... even again more recently.

The fact they've refurbished their site's look several times this year, but neglected to include more useful settings, suggests they don't give a damn.

**Comment by methodbox** [comment]
The reason is that the coding standards vary between organizations.

There is not necessarily a one size fits all solution, so allowing it to be customized is the best alternative.

The problem is that employing an account setting for this would involve a lot of backward configuration consideration to support accounts that didn't have a .editorconfig, but adding one to each repo would also mean that everyone could customize it as they see fit.

Last, the problem with requiring a pull because a single file was created for people who didn't necessarily ask for that file would likely piss off a lot of users, yet using the config file is the only way they could make this setting customizable.

Alternatively, they could define this in CSS, but then we have the problem of overriding his with non-standard spacing like 3 or 5, etc.

There's not an elegant solution.

**Comment by aristidesfl** [comment]
Are there any editors which come pre-configured with tabs=8 spaces?

**Comment by methodbox** [comment]
@aristidesfl

Most default to 4, but can easily be edited.

For Sublime, this can be set in user preferences.

Just google, "sublime tab spacing".

**Comment by aristidesfl** [comment]
> For Sublime, this can be set in user preferences.
> Just google, "sublime tab spacing".

![](http://s2.quickmeme.com/img/1b/1b5ce691ca1447e4a3b3af4ae665366ea4bcfa5363eb7b7b62e7ae2aec09cb68.jpg)

> Most default to 4, but can easily be edited.

I'm asking about examples of editors which default to **8**

**Comment by methodbox** [comment]
Yeah, I know, dick.

Just set it to 8.

No editors default to 8 because nobody uses that as a coding standard.

**Comment by aristidesfl** [comment]
If not editors use 8 as default, why would ACE (or github) use that as default?

**Comment by WORMSS** [comment]
Easier to put your hands up and say `"We use the browser default"` and piss both `2 space` tabs and `4 space` tab users than to take sides.

**Comment by aristidesfl** [comment]
Anyone knows why 8 is the browser default?

**Comment by tinco** [comment]
It is defined here: https://www.w3.org/TR/CSS21/text.html#white-space-model and later redefined here: https://drafts.csswg.org/css-text-3/#tab-size-property. As to why it was defined that way.. no idea.

**Comment by anhlqn** [comment]
What if I want indent style to be spaces instead of tabs? Would this work?
```
root = true

[*]
indent_style = space
indent_size = 4
```

**Comment by drydenp** [comment]
So happy this works. I would prefer a setting on GitHub and actually I would prefer the default to simply be changed to 4, because it seems ridiculous that *any* project depends on this being 8, why does Ruby/ruby have this as their config?

    tab_width = 8
    indent_style = tab
    indent_size = 4

So they use an indent_size of 4, but a tab_width of 8. And indeed, files like Makefiles are displayed using 8 spaces. So GitHub apparently honours the tab_width parameter over indent_size in its display on GitHub for ruby/ruby. This means Ruby/ruby won't be hurt by setting the default to 4, because they've already overridden it with 8. Otherwise, it would be displayed as 4 seeing as indent_size would take effect.

So, the argument that ruby/ruby uses a size of 8 is already devoid because .editorconfig is now honoured and honoured well. This means that any actual projects that really care about this issue are already going to have a .editorconfig defined and they won't be hurt by a new default of 4. This leaves the door wide open for setting a better default, which I argue would be 4.

Particularly of course Ruby doesn't even use tabs in their main files.

So as others have said, a space-oriented project would dictate rules for tab-oriented files. Because it happens to use a few of them...

> What if I want indent style to be spaces instead of tabs? Would this work?

It would have no effect on GitHub as you are already using spaces which won't be changed. I guess the intent is for an editor to change your tabs to 4 spaces when you use a tab in your editor. But it won't affect display, only tab conversion in your editor.

**Comment by ioquatix** [comment]
I added a .editorconfig and while browsing code in `lib/` was affected, code in `README.md` was not.

https://github.com/socketry/async

Any ideas how to get tab-width set correctly in `README.md`?

**Comment by Alhadis** [comment]
@ioquatix This appears to be GitHub's fault, not yours. You should report this to their support team, and/or file an issue at [`github/markup`](https://github.com/github/markup), which I believe is responsible for rendering markdown into the HTML served to readers.

**Comment by ioquatix** [comment]
@Alhadis I send them an email about it.

> I believe that you should add some new preferences:
> 
> 1/ A per-user/organisation tab-width preference which applies to all code within that user/organisation.
> 
> 2/ A per-project tab-width preference, which overrides the previously mentioned one.
> 
> 3/ And finally, a per-user tab-width preference which overrides 1/ above but not 2/, which applies to code the user is viewing.
> 
> 4/ If an .editorconfig file exists, it overrides everything for that project.
> 
> The current state w.r.t. tabs in source code on GitHub is pretty bad, tabs with width = 8 is not ideal.
> 
> Additionally, 3/ above could have a min/max for mobile/desktop viewing, respectively - this is very nice because tab width can be 2 on a mobile browser and 4 on a desktop.
> 
> Kind regards,
> Samuel

Already got a reply:

> Hi Samuel,
> 
> Thank you for taking the time to write such a detailed and well-reasoned feature request. This sounds like a great idea, and while I can't make any promises, I'm definitely going to pass this idea on to the team for their consideration.
> 
> Please let us know if there's anything else we can do for you in the meantime, and have a wonderful day!
> 
> All the best,
> Nick
> @nickcannariato
> GitHub Support

**Comment by josduj** [comment]
I don't know if that was intentional, but it seems that `.editorconfig` doesn't work for hidden files.

**Comment by Alhadis** [comment]
What exactly are you referring to when you say "hidden files"...?

**Comment by josduj** [comment]
@Alhadis dot-files

**Comment by ioquatix** [comment]
The update I received here is that this is not currently a priority: https://github.com/github/markup/issues/1045#issuecomment-327074441 Just FYI.

**Comment by iyerusad** [comment]
@josduj I found the following would work for dotfiles:
```
root = true

[*]
indent_style = tab 
indent_size = 2 

[*.md]
indent_style = tab 
indent_size = 2 

[.*]
indent_style = tab 
indent_size = 2 
```

**Comment by HaraldNordgren** [comment]
I don't this `root` stuff is necessary, so this should be enough for your Go repos:

    [*.go]
    indent_size = 4

**Comment by ClaudeDev** [comment]
I was sure it was an implemented feature long ago!
We need to able to select tabs or spaces in the settings.
It is fairly simple, right?

**Comment by LorenzoTa** [comment]
having the possibility to choice this among settings will be good; per repository will be optimal since some language seems to be affected by indentation.

It's very annoying pasting code and see the indentation becoming eccesive.

I think we all wait for this feature with patience.

**Comment by amalic** [comment]
Tab width setting would be awesome!

**Comment by s-h-a-d-o-w** [comment]
One more workaround (a Userscript) for people who like me prefer something browser-independent that's integrated with the GitHub website: https://github.com/s-h-a-d-o-w/github-userscripts

**Comment by h-h-h-h** [comment]
@Rich-Harris:

> browsers render tabs as 8 spaces. Even among the minority of developers who agree that tabs > spaces, I've yet to find anyone who thinks that makes sense.

Just FYI:

> Coding Style
> ============
> ...
> - Yosys code is using tabs for indentation. Tabs are 8 characters.

Source: https://github.com/YosysHQ/yosys/blob/master/CodingReadme#L221

---

@franciscolourenco:

> Are there any editors which come pre-configured with tabs=8 spaces?

Yes, Windows' Notepad. :)

---

<sup>PS: To be clear: I'm no fan of a tab width of 8 spaces.</sup>

**Comment by drydenp** [comment]
There is actually someone who uses an indent of 8.

But they are spaces.

You will find such hideosity when you visit.....

**Comment by maximal** [comment]
I think it should not be constant (as described here) and repository level configuration would be just fine.

Created an issue for configurable setting: #1490. If you feel it like that, vote there please.

**Comment by guruskid** [comment]
I am building a project but am looking for a full-stack dev that will join me on the project, if anyone is interested should contact me on configureall@gmail.com with the title PROJECT

**Comment by WORMSS** [comment]
@guruskid you can get reported for spamming, and have your github account made invisible. It would make all your github issues/pull requests/repos/comments all invisible to the outside world, but you are none the wiser. Highly recommend leaving recruiting to recruiting websites.

**Comment by thejpster** [comment]
Indentation depth should be a per-user option, not a per-project option. There are many visual impairment reasons to prefer a very shallow indent (if you need a large font size), or a very deep indent (if it makes it easier to see the indentation).

I would like to be able to set my default tab size in my profile, and then if we all use tabs for indentation, we all get the indentation depth that suits us.

**Comment by Alhadis** [comment]
@thejpster I agree, but that wouldn't help users without GitHub accounts (or who are simply reading code whilst logged out).

A per-project option is at least a step in the right direction.

**Comment by ioquatix** [comment]
In theory, someone can set their own CSS rules in their user agent. Something like:

```
pre {
  tab-size: 4;
}
```

to your own user agent might be an option for the "per-user" option?

Not that I don't think GitHub could do a better job in this area.

**Comment by Alhadis** [comment]
@ioquatix What about mobile users, or those whose browsers don't support such a thing?

**Comment by ioquatix** [comment]
Oh, I totally agree, and I advocate for:

- Per project setting (some projects using hard tabs simply require them to be 8 spaces, they are not a form of variable indentation).
- Per organisation default (if nothing else is configured, use the specified default).
- Per user default (if this is set, it overrides per-organisation defaults).

**Comment by natematykiewicz** [comment]
@ioquatix Ironically, I found this post by reading some of your code and wishing I could make the tabs be 2 spaces. 😅

**Comment by Alhadis** [comment]
It's time for my annual *"nag GitHub support to add this feature again*". This time I was more clear about receiving an explanation. CC'd here for transparency:

> Dear GitHub,
>
> I've requested this almost every year [since 2015](https://github.com/isaacs/github/issues/170#issuecomment-135785607), and will continue to do so: **_Please add a repository-level setting to change the default tab-width of `8` columns._** You added [`.editorconfig` support](https://github.com/isaacs/github/issues/170#issuecomment-150489692), but that isn't an adequate solution — you're forcing users to abuse an EditorConfig setting that affects how collaborators can contribute to their repository.
>
> The **whole point** of semantic tab-use is to enable others to read your code with the indentation-size they're most comfortable with. This might be simple preference, or a product of physical constraints such as screen- or font-size. By requiring `indent_size = 4` to override an unwieldy historical default, you're taking this benefit away from users whose codebases are formatted with accessibility in mind.
>
> ***PLEASE*** add a repository-level setting for this. It's a simple case of adding the following CSS to a page's header:
>
> ~~~html
> <style>
> 	/* Replace `8` with configured repository setting */
> 	:root{
> 		--tab-size: 8;
> 	}
> 
> 	/* Add any other selectors styled with `white-space: pre;` */
> 	.blob-code-inner, pre, code, tt {
> 		-moz-tab-size: var(--tab-size);
> 		tab-size:      var(--tab-size);
> 	}
> </style>
> ~~~
>
> The solution above uses a [CSS variable](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) that can be overridden in specific elements by adding `style="--tab-size: 4"`, if such a thing is needed.
>
> #### If there's a reason GitHub staff refuse to add this feature, please e-mail me the official explanation.
>
> Punting this as "low priority" is absurd when it takes very little effort to add, and *many* users have [requested it](https://github.com/isaacs/github/issues/170) time and time again… for years.
>
> Moreover, [you listened to my incessant nagging](https://github.com/github/linguist/issues/4291#issuecomment-452751949) about adding 1-pixel dividers between language-colours in repository stat-bars, which is no less complicated (and far more trivial) than adding a tab-size feature.
>
> Yours,  
> ⸻ @Alhadis

**Comment by ioquatix** [comment]
One way to get GitHub to move on this might be to implement a similar feature in GitLab...

https://gitlab.com/gitlab-org/gitlab/issues/26768

I'm going to take a stab at it if I have time.

**Comment by maximal** [comment]
> One way to get GitHub to move on this might be to implement a similar feature in GitLab...
>
> https://gitlab.com/gitlab-org/gitlab/issues/26768

Yes, guys, comment and upvote there!

**Comment by melgu** [comment]
This also applies to Markdown previews.
For those not even the `?ts=4` trick works.

**Comment by dschinkel** [comment]
yea this whole thing is stupid that we can't edit the Gists to make the tabs and spaces show up how we'd prefer. 

Maybe it’s because Microsoft owns GitHub know and does stupid crap like this? Who the hell wants it 8 spaces. People usually code in 2 or 4 so this should be configurable so that anyone looking at our gists sees what we want them to see and ourselves.

This reminds me of the half assed products MS comes out with. They don’t care and they leave things half baked just like products like MS Teams etc. People want to be able to configur this stuff, not add hacks to the URL to “make it formatted”

The problem with MS has always been even when I started coding professionally in 2000 is that they don’t listen to customer feedback. What they implement is what only they think people need. Nothing has changed with this organization. Looks like this is still the case with the inability to paste in and specify tabs or spaces configured for viewing Gists as the code should be viewed.

**Comment by RomainLanz** [comment]
Any news on this feature?

**Comment by Primajin** [comment]
Since Github offers things like setting emoji color and theme in here: https://github.com/settings/appearance this should be the right place to also set preferred tab size.

**Comment by botatoes** [comment]
Hi all! Was wondering if configuring tab sizes on the editor here resolves this issue.  
<img width="326" alt="image" src="https://user-images.githubusercontent.com/7900087/116297281-f6373400-a74f-11eb-86c5-b22904b61043.png">
<img width="286" alt="image" src="https://user-images.githubusercontent.com/7900087/116297509-2e3e7700-a750-11eb-821c-ef8c008dd8c8.png">

**Comment by AverageHelper** [comment]
Since I just found this, and I don't know if it's been posted already, and I think it makes relevant arguments, I'll link [this Reddit post](https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/).

TL;DR:
There is a significant population of developers and potential developers with visual impairments who rely on editor-level tab-size configuration to do their work. The fact that GitHub et al do not permit this sort of configuration is a pain point for them.

+1 for adding an optional user-level tab-width config. Bonus points to actively recommend that new repos use tabs.

**Comment by Rybec** [comment]
I just want to express my desire for a repo level tab setting that does not require pushing contributors to use that same setting in their own editors.  I go to great effort to make sure my code looks right in any tab setting (tabs for indentation, spaces for alignment), but I want the Github presentation to be 4 character widths for tabs.  Tabs more than 4 characters wide rapidly push code so far to the right that it can become a serious problem for readability.

**Comment by GregWoods** [comment]
just created my first gist. Yay!
But wait, what the hell is wrong with formatting?
Ah wait there's an open issue for this.
But it is 7 years old
Goodbye gists :(

**Comment by Levi-Lesches** [comment]
I didn't realize there wasn't anything open at github/feedback, so I opened https://github.com/github/feedback/discussions/4893 and linked back here. Upvote (not just 👍) there to get GitHub's attention on this.

**Comment by michellemerrill** [comment]
Hey everyone 👋 good news, you can now define the tab size in the [Appearance settings](https://github.com/settings/appearance) of your user account. Code rendered on GitHub with tab indent will use your preferred tab size.
